### ISSUE TITLE
Where is safe_run_with_state ?

### ISSUE THREAD

**Issue by sgrimee** [Ekranos/nagiosplugin-rs#7]
I am trying to implement a simple behaviour, were all errors would return as state Unknown instead of Critical.
```rust
    Runner::new()
        .safe_run_with_state(do_check, ServiceState::Unknown)
        .print_and_exit();
```
However this gives:
```rust
error[E0599]: no method named `safe_run_with_state` found for struct `Runner` in the current scope
  --> src/main.rs:47:10
   |
47 |         .safe_run_with_state(do_check, ServiceState::Unknown)
   |          ^^^^^^^^^^^^^^^^^^^ method not found in `Runner<_>`
```
It seems the ```safe_run_with_state``` helper and the helper.rs file have disappeared altogether (using ```nagiosplugin = "0.4"```).
I have tried doing it with on_error but was not able to find the right syntax. Any help welcome!

**Comment by vbrandl** [comment]
I'm not 100% sure but the test code might help you: https://github.com/Ekranos/nagiosplugin-rs/blob/master/src/runner.rs#L83

**Comment by sgrimee** [comment]
thanks @vbrandl . Yes I suspect this is the new way of doing it, but I was not able to find the right syntax to just propagate the error and specify the State.
```rust
    Runner::new()
        .on_error(|e| (ServiceState::Unknown, e))
        .safe_run(do_check)
        .print_and_exit();
```
gives
```rust
error[E0308]: mismatched types
  --> src/main.rs:47:47
   |
47 |         .on_error(|e| (ServiceState::Unknown, e))
   |                                               ^- help: try using a conversion method: `.to_string()`
   |                                               |
   |                                               cyclic type of infinite size
```
Clearly I largely don't know what I am doing, any pointer welcome ;-)

**Comment by vbrandl** [comment]
So your `do_check` function returns `Result<Resource, E>`, `E` being your custom error type.

If you take a look at the parameters of `on_error`, you see, that the function passed to `on_error` takes a reference `&E` but should return a tuple of `(ServiceState, E)` (notice that in the return type, `E` is owned and not borrowed):

```rust
pub fn on_error(mut self, f: impl FnOnce(&E) -> (ServiceState, E) + 'static) -> Self
```

The easiest solution would be to make your error type `Clone` or even `Copy`:


```rust
use nagiosplugin::{Resource, Runner, ServiceState};

#[derive(Copy, Clone, Debug)]
enum MyError {
    UnknownErrorType,
}

fn noop() -> Result<Resource, MyError> {
    Err(MyError::UnknownErrorType)
}

fn main() {
    Runner::<MyError>::new()
        .on_error(|e| (ServiceState::Unknown, *e))
        // if your error type does not implement `Copy`, use the following line:
        //         .on_error(|e| (ServiceState::Unknown, e.clone()))
        .safe_run(noop)
        .print_and_exit();
}
```

**Comment by sgrimee** [comment]
You are right, I missed that ownership transfer. I am not using a custom Error type, these are anyhow::Error, which is not Clone AFAIK. I got it to work the way like this, may not be optimal but does the trick:
```rust
    Runner::new()
        .on_error(|e| -> (ServiceState, anyhow::Error) {
            (ServiceState::Unknown, anyhow!(e.to_string()))
        })
        .safe_run(do_check)
        .print_and_exit();
```
So thanks for the help! I think it would be awesome if some documentation would give examples for this use case, as in #2, where I got the outdated reference for ```safe_run_with_state```.

**Comment by Ekranos** [comment]
I'll try to get up an example with anyhow up this week. In case it makes sense I will also update the library to provide some way to just map to another ServiceState on error.

That there is no easy and straight forward way to just remap the ServiceState is an oversight from me, I am sorry about that.

I will close this issue with the coming commit.

**Comment by vbrandl** [comment]
Is there a reason why the function passed to `on_error` isn't simply changed to `impl FnOnce(E) -> (ServiceState, E) + 'static`? This should make working with the `on_error` mapper easier since there is no need for `E` to be `Copy` or `Clone`.

Changing the function signature as described and changing the `safe_run` function like below seems to work (e.g. compile and pass the tests) fine for me.

```rust
    pub fn safe_run(self, f: impl FnOnce() -> Result<Resource, E>) -> RunnerResult<E> {
        match f() {
            Ok(resource) => RunnerResult::Ok(resource),
            Err(err) => {
                let (state, msg) = if let Some(error_mapper) = self.on_error {
                    error_mapper(err)
                } else {
                    (ServiceState::Critical, err)
                };

                RunnerResult::Err(state, msg)
            }
        }
    }
```

**Comment by Ekranos** [comment]
I resolved this problem in 1557e59a96b3a2526c2e8016f38b650977489f58. Just use version `0.5.*`.

See the example: https://github.com/Ekranos/nagiosplugin-rs/blob/1557e59a96b3a2526c2e8016f38b650977489f58/examples/change_error_state.rs

In case I missed the point of this issue, please reopen it.

**Comment by sgrimee** [comment]
thank you @Ekranos. Much cleaner. I have updated my code in check_brandmeister.

**Comment by Ekranos** [comment]
@sgrimee in https://github.com/sgrimee/check_brandmeister/blob/master/src/bin/check_brandmeister.rs#L84 will never trigger. This is exclusively for the minimum and maximum fields for the performance data. You can read about it here https://nagios-plugins.org/doc/guidelines.html#AEN200

The format in the end is `'label'=value[UOM];[warn];[crit];[min];[max]`. As far as I know, Icinga/Nagios forward these to Graphite or Influx etc. and you can use these values in your graphs in Grafana or Graphite. In your case you could set a minimum of 0 minutes, cause you can't see into the future ;).

You may also want to take a look at version 0.5.1 where you can now use only a warning or a critical threshold, in case you want only warning or critical. This way users of your check could decide if they only want a critical or warning or maybe even nothing at all and just use the check for data collection. You should only need to change https://github.com/sgrimee/check_brandmeister/blob/0e49dddd973f78c7041346fa382112d4affbf186/src/bin/check_brandmeister.rs#L11 to `Option<i64>` and remove the `default_value` on https://github.com/sgrimee/check_brandmeister/blob/0e49dddd973f78c7041346fa382112d4affbf186/src/bin/check_brandmeister.rs#L37 and https://github.com/sgrimee/check_brandmeister/blob/0e49dddd973f78c7041346fa382112d4affbf186/src/bin/check_brandmeister.rs#L47 and it should **just work**.

I am rambling too much, sorry for that. Have a nice day!

**Comment by Ekranos** [comment]
Also I saw you forked a fork which re-added unit of measurement. I also forgot them when I redid the lib. They are now back in version 0.5.2. See the example: https://github.com/Ekranos/nagiosplugin-rs/blob/accd9fabcc6228a71b410bc64cb71b1bb7a287fa/examples/change_error_state.rs

**Comment by sgrimee** [comment]
@Ekranos thanks a lot for checking my code and for your advice!
Yes, I have realised only recently that the maximum and minimum methods are just for nagios and not meant to set a threshold. Somehow my git push this morning did not go through, now the code has this comment removed. Also the fork was precisely to troubleshoot this, I wanted to help with a fix but it is not needed ;-) I have deleted the fork.
I will look into the new warning only and critical only, cheers!
